lib/SPIRV/libSPIRV/SPIRVValue.cpp

```diff
@@ -121,4 +124,36 @@ void SPIRVValue::setFPFastMathMode(SPIRVWord M) {
                      << "\n")
 }
 
+template <spv::Op OC>
+void SPIRVConstantBase<OC>::setWords(const uint64_t *TheValue) {
+  assert(TheValue && "Nullptr value");
+  recalculateWordCount();
+  validate();
+
+  Words.resize(NumWords);
+  for (size_t I = 0; I != NumWords / 2; ++I) {
+    Words[I * 2] = static_cast<SPIRVWord>(TheValue[I]) & SPIRVWORD_MAX;
+    Words[I * 2 + 1] =
+        static_cast<SPIRVWord>((TheValue[I] >> SpirvWordBitWidth)) &
+        SPIRVWORD_MAX;
+  }
+  if (NumWords % 2)
+    Words.back() =
+        static_cast<SPIRVWord>(TheValue[NumWords / 2]) & SPIRVWORD_MAX;
+}
+
+// Complete constructor for AP integer constant
+template <spv::Op OC>
+SPIRVConstantBase<OC>::SPIRVConstantBase(SPIRVModule *M, SPIRVType *TheType,
+                                         SPIRVId TheId,
+                                         const llvm::APInt &TheValue)
+    : SPIRVValue(M, 0, OC, TheType, TheId) {
+  setWords(TheValue.getRawData());
+}
+
+// To solve errors about undefined reference to template class methods
+// definitions.
+template class SPIRVConstantBase<OpConstant>;
+template class SPIRVConstantBase<OpSpecConstant>;
```

Comment:
Do we still need these lines?

Reply:
Yes. Otherwise, compilation of libLLVMSPIRVLib.a is failed with 'undefined reference' error for setValue method.